pkg/alertmanager/distributor.go

```diff
@@ -0,0 +1,235 @@
+
+	if r.Method == http.MethodGet || r.Method == http.MethodHead {
+		d.doRead(userID, w, r, logger)
+	} else {
+		d.doWrite(userID, w, r, logger)
+	}
+}
+
+func (d *Distributor) doWrite(userID string, w http.ResponseWriter, r *http.Request, logger log.Logger) {
+	var body []byte
+	var err error
+	if r.Body != nil {
+		body, err = ioutil.ReadAll(r.Body)
+		if err != nil {
+			if errors.Is(util.ErrRequestBodyTooLarge, err) {
+				http.Error(w, "Request body too large", http.StatusRequestEntityTooLarge)
+			}
+			level.Error(logger).Log("msg", "failed to read the request body during write", "err", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+
+	var firstSuccessfulResponse *httpgrpc.HTTPResponse
+	var firstSuccessfulResponseMtx sync.Mutex
+	grpcHeaders := httpTogrpchttpHeaders(r.Header)
+	err = ring.DoBatch(r.Context(), RingOp, d.alertmanagerRing, []uint32{shardByUser(userID)}, func(am ring.InstanceDesc, _ []int) error {
+		// Use a background context to make sure all alertmanagers get the request even if we return early.
+		localCtx := user.InjectOrgID(context.Background(), userID)
+		sp, localCtx := opentracing.StartSpanFromContext(localCtx, "Distribute.doWrite")
+		defer sp.Finish()
+
+		resp, err := d.doRequest(localCtx, am, &httpgrpc.HTTPRequest{
+			Method:  r.Method,
+			Url:     r.RequestURI,
+			Body:    body,
+			Headers: grpcHeaders,
+		})
+		if err != nil {
+			return err
+		}
+
+		if resp.Code/100 != 2 {
+			return httpgrpc.ErrorFromHTTPResponse(resp)
+		}
+
+		firstSuccessfulResponseMtx.Lock()
+		if firstSuccessfulResponse == nil {
+			firstSuccessfulResponse = resp
+		}
+		firstSuccessfulResponseMtx.Unlock()
+
+		return nil
+	}, func() {})
+
+	if err != nil {
+		respondFromError(err, w, logger)
+	}
+
+	firstSuccessfulResponseMtx.Lock() // Another request might be ongoing after quorum.
+	resp := firstSuccessfulResponse
+	firstSuccessfulResponseMtx.Unlock()
+
+	if resp != nil {
+		respondFromHTTPGRPCResponse(w, resp)
+	} else {
+		// This should not happen.
+		level.Error(logger).Log("msg", "distributor did not receive response from alertmanager though no errors")
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+}
+
+func (d *Distributor) doRead(userID string, w http.ResponseWriter, r *http.Request, logger log.Logger) {
+	key := shardByUser(userID)
+	replicationSet, err := d.alertmanagerRing.Get(key, RingOp, nil, nil, nil)
+	if err != nil {
+		level.Error(logger).Log("msg", "failed to get replication set from the ring", "err", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	req, err := server.HTTPRequest(r)
+	if err != nil {
+		level.Error(logger).Log("msg", "failed to get grpc request from http request", "err", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	sp, ctx := opentracing.StartSpanFromContext(r.Context(), "Distribute.doRead")
+	defer sp.Finish()
+	// Until we have a mechanism to combine the results from multiple alertmanagers,
```

Comment:
What would combining mean in terms of behavior?

Reply:
Combining would deduplicate the results from those alertmanagers, basically a union